### DISCUSSION TITLE
Invalid prettier configuration file detected

### DISCUSSION THREAD

**Discussion by mxz7** [prettier/prettier#15167] [Q&A]
<img width="837" alt="image" src="https://github.com/prettier/prettier/assets/29782397/f1ec3d3c-9147-4262-8e35-b9da0432cdb4">

<img width="843" alt="image" src="https://github.com/prettier/prettier/assets/29782397/b33baca7-d7aa-45d6-a7f3-b983b40bbdac">

I'm properly baffled. it used to work a few weeks ago

**Comment by kachkaev** [comment]
Hi @tekoh! It’s hard to help because it’s not clear what version of Prettier you are using, what is you `module` attribute in `package.json` and what’s the name of the config file.

If you use Prettier v3, you ca try removing `require` wrapper calls and see if it helps. See https://prettier.io/docs/en/configuration.html for more info

**Reply by mxz7** [reply]
prettier v2.8.8

there is no module attribute in my package.json but type is set to module

config file is .prettierrc.cjs

**Reply by thomo** [reply]
Same issue here:
- VSCode project running in a devcontainer
- prettier version 3.0.0
- config .prettierrc.cjs
  ```
  const config = {
      plugins: [require("prettier-plugin-tailwindcss")],
  }
  
  module.exports = config;
  ```
- no prettier section in package.json

At start all looks ok - but when I open a file and execute "Format file" the error occurs:

Log Output:
```
["INFO" - 9:07:19 AM] Extension Name: esbenp.prettier-vscode.
["INFO" - 9:07:19 AM] Extension Version: 10.1.0.
["DEBUG" - 9:07:19 AM] Enabling Prettier globally
{
  "languageSelector": [
    {
      "language": "javascript"
    },
    {
      "language": "mongo"
    },
    {
      "language": "javascriptreact"
    },
    {
      "language": "typescript"
    },
    {
      "language": "typescriptreact"
    },
    {
      "language": "json"
    },
    {
      "language": "jsonc"
    },
    {
      "language": "json5"
    },
    {
      "language": "css"
    },
    {
      "language": "postcss"
    },
    {
      "language": "less"
    },
    {
      "language": "scss"
    },
    {
      "language": "handlebars"
    },
    {
      "language": "graphql"
    },
    {
      "language": "markdown"
    },
    {
      "language": "mdx"
    },
    {
      "language": "html"
    },
    {
      "language": "vue"
    },
    {
      "language": "yaml"
    },
    {
      "language": "ansible"
    },
    {
      "language": "home-assistant"
    },
    {
      "language": "jsonc",
      "scheme": "vscode-userdata"
    }
  ],
  "rangeLanguageSelector": [
    {
      "language": "javascript"
    },
    {
      "language": "javascriptreact"
    },
    {
      "language": "typescript"
    },
    {
      "language": "typescriptreact"
    },
    {
      "language": "json"
    },
    {
      "language": "jsonc"
    },
    {
      "language": "graphql"
    }
  ]
}
["DEBUG" - 9:07:34 AM] Local prettier module path: '/workspaces/logbook-ui/node_modules/prettier/index.cjs'
["DEBUG" - 9:07:34 AM] Using prettier version 3.0.0
["ERROR" - 9:07:34 AM] Invalid prettier configuration file detected.
{}
["ERROR" - 9:07:34 AM] Invalid prettier configuration file detected. See log for details.
["DEBUG" - 9:07:34 AM] Enabling Prettier for Workspace /workspaces/logbook-ui
{
  "languageSelector": [
    {
      "pattern": "/workspaces/logbook-ui/**/*.{js,_js,bones,cjs,es,es6,frag,gs,jake,javascript,jsb,jscad,jsfl,jslib,jsm,jspre,jss,mjs,njs,pac,sjs,ssjs,xsjs,xsjslib,wxs,js.flow,jsx,ts,cts,mts,tsx,importmap,json,4DForm,4DProject,avsc,geojson,gltf,har,ice,JSON-tmLanguage,mcmeta,tfstate,tfstate.backup,topojson,webapp,webmanifest,yy,yyp,jsonc,code-snippets,sublime-build,sublime-commands,sublime-completions,sublime-keymap,sublime-macro,sublime-menu,sublime-mousemap,sublime-project,sublime-settings,sublime-theme,sublime-workspace,sublime_metrics,sublime_session,json5,css,wxss,pcss,postcss,less,scss,handlebars,hbs,graphql,gql,graphqls,md,livemd,markdown,mdown,mdwn,mkd,mkdn,mkdown,ronn,scd,workbook,mdx,component.html,html,hta,htm,html.hl,inc,xht,xhtml,mjml,vue,yml,mir,reek,rviz,sublime-syntax,syntax,yaml,yaml-tmlanguage,yaml.sed,yml.mysql}",
      "scheme": "file"
    },
    {
      "language": "javascript"
    },
    {
      "language": "mongo"
    },
    {
      "language": "javascriptreact"
    },
    {
      "language": "typescript"
    },
    {
      "language": "typescriptreact"
    },
    {
      "language": "json"
    },
    {
      "language": "jsonc"
    },
    {
      "language": "json5"
    },
    {
      "language": "css"
    },
    {
      "language": "postcss"
    },
    {
      "language": "less"
    },
    {
      "language": "scss"
    },
    {
      "language": "handlebars"
    },
    {
      "language": "graphql"
    },
    {
      "language": "markdown"
    },
    {
      "language": "mdx"
    },
    {
      "language": "html"
    },
    {
      "language": "vue"
    },
    {
      "language": "yaml"
    },
    {
      "language": "ansible"
    },
    {
      "language": "home-assistant"
    },
    {
      "language": "jsonc",
      "scheme": "vscode-userdata"
    }
  ],
  "rangeLanguageSelector": [
    {
      "language": "javascript"
    },
    {
      "language": "javascriptreact"
    },
    {
      "language": "typescript"
    },
    {
      "language": "typescriptreact"
    },
    {
      "language": "json"
    },
    {
      "language": "jsonc"
    },
    {
      "language": "graphql"
    }
  ]
}
["INFO" - 9:07:43 AM] Formatting file:///workspaces/logbook-ui/src/App.tsx
["DEBUG" - 9:07:43 AM] Local prettier module path: '/workspaces/logbook-ui/node_modules/prettier/index.cjs'
["DEBUG" - 9:07:43 AM] Using prettier version 3.0.0
["ERROR" - 9:07:43 AM] Invalid prettier configuration file detected.
{}
["ERROR" - 9:07:43 AM] Invalid prettier configuration file detected. See log for details.
```

**Reply by SapphoSys** [reply]
@kachkaev Hi there! I ran into the same issue.

This is the error I'm experiencing:

```
["INFO" - 3:13:51 PM] Formatting file:///home/chloe/projects/site/prettier.config.mjs
["ERROR" - 3:13:51 PM] Invalid prettier configuration file detected.
{}
["ERROR" - 3:13:51 PM] Invalid prettier configuration file detected. See log for details.
```

Here's my Prettier configuration file:
`prettier.config.mjs`
```js
/** @type import('prettier').Config */
export default {
  plugins: [
    "prettier-plugin-astro",
    "prettier-plugin-tailwindcss",
  ],
  proseWrap: "always",
  overrides: [
    {
      files: "*.astro",
      options: { parser: "astro" },
    },
  ],
};
```

Specifications:
OS: Ubuntu 22.04.2 LTS under WSL2 (Windows 11)
Prettier version: 
```
$ yarn prettier -v
3.0.2
```
Prettier plugin versions:
`package.json`
```
"prettier-plugin-astro": "^0.11.1",
"prettier-plugin-tailwindcss": "^0.5.3"
```

**Comment by electrovir** [comment]
This happens to me when I switch from `prettier.config.cjs` to `prettier.config.mjs`. It seems that switching to `.mjs` triggers Prettier to enter some kind of ESM mode which then makes it incapable of formatting CommonJS `.js` files: they must all be renamed to `.cjs`.

**Comment by ethanwu10** [comment]
The problem seems to be that in the VSCode extension, module objects can't be passed through the RPC boundary between the main extension and the worker process where config is actually read. For the VSCode extension to work, the Prettier config must be effectively plain old data (specifically, the [HTML structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm)) due to [restrictions in `worker_threads`](https://nodejs.org/api/worker_threads.html#portpostmessagevalue-transferlist) from the way the VSCode extension is implemented.

I ran into this while trying to get a plugin to load under PNP, and it seems like there is no way to make the VSCode extension work. Doing the import in the config doesn't work because of this issue. Meanwhile, only resolving the import in the config (via `import.meta.resolve`) also doesn't work since the VSCode main thread then tries to resolve it again using `require.resolve`: `import.meta.resolve` gives a URI to a ESM file, which the CJS `require` machinery seems to reject.

**Comment by trevorr** [comment]
I ran into this as well and "fixed" it by switching to a JSON configuration.

**Comment by Relequestual** [comment]
I had this issue in VSCode. I eventually solved it. Here are some learnings.

> If you use Prettier v3, you ca try removing require wrapper calls and see if it helps

It did help.

- The Prettier config file you used was valid.
- The Prettier VS Code extension can sometimes cache errors or not pick up config changes until you reload the window (Developer: Reload Window).
- "Format on Save" must be enabled in VS Code settings for automatic formatting to work.
- Using "Format Document" from the command palette can help test if Prettier is working, even if "Format on Save" is off.
- After changing Prettier config files, always reload the VS Code window to ensure the extension picks up the changes.

Hopefully this helps others. It can be so frustrating to get no answers and run out of things to try.
